esx_service/utils/auth_api.py

```diff
@@ -569,15 +569,21 @@ def _tenant_rm(name, remove_volumes=False):
         error_info = generate_error_info(ErrorCode.TENANT_NOT_EXIST, name)
         return error_info
 
-    if tenant.vms:
-        error_info = generate_error_info(ErrorCode.TENANT_NOT_EMPTY, name)
-        logging.error(error_info.msg)
-        return error_info
-
     error_info, auth_mgr = get_auth_mgr_object()
     if error_info:
         return error_info
 
+    if tenant.vms:
+        if not force:
+            error_info = generate_error_info(ErrorCode.TENANT_NOT_EMPTY, name)
+            logging.error(error_info.msg)
+            return error_info
+        else:
+            error_msg = tenant.remove_vms(auth_mgr.conn, tenant.vms)
```

Comment:
Isn't this call going to return some error due to "Cannot find vm"?

Reply:
This call is the final step to removing the VM from the vmgroup, the vms list is returned from _get_tenant()_ which returns all the VMs associated with the tenant as (vm_id, vm_name).  tenant.remove_vms() only removes the vms from the DB, any error at that point is treated as _internal error_. No checks are performed on the vm IDs,

Reply:
I still don't follow. My question is: if the VMs were already deleted, doesn't the call tenant.remove_vms() return some error which will again fail the "vmgroup rm" operation?

Reply:
As per offline discussion, in this case the VMs are removed from VC, but still exist in vDVS database. So above tenant.remove_vms() call will not fail.